Java support

alshain opened the pull request:
![image](https://user-images.githubusercontent.com/349397/58385239-7ba68c80-7fed-11e9-830c-90f4e527acda.png)

This adds Java support to compiler explorer by compiling through `javac` and using `javap` to disassemble.

Compared to most other compiled languages, Java generates one class file per class. the `objdump` method will simply read all all `*.class` inside the directory and disassembles them.

Also see: #12 #1364 #1058

mattgodbolt commented:
Thanks for this! We'll take a look! We will need a way of installing `javac` and `javap` in the https://github.com/mattgodbolt/compiler-explorer-image repository too (there are scripts in `update_compilers/*.sh` that will need to be modified. We can help with that, of course :)

mattgodbolt commented:
See #1058

mattgodbolt commented:
Apologies for being slow: we're trying to get the scary big "execution support" pushed to prod. Once that's done we will look to merge and release this! Thanks for your patience.

RabsRincon commented on lib/compilers/java.js:
```diff
@@ -0,0 +1,265 @@
+// Copyright (c) 2017, Rubén Rincón
```
I'm flattered, but the copyright is not right here :P
If you wish, you can change it to
```diff
- // Copyright (c) 2017, Rubén Rincón
+ // Copyright (c) 2019, Compiler Explorer Authors
```

alshain replied:
:) changed it

RabsRincon commented on lib/compilers/java.js:
```diff
@@ -0,0 +1,265 @@
+        let toSkip = 0;
+
+        const oneArgBlacklist = [
+            // -d directory
```
Thanks for explaining all the different flags! :)

RabsRincon commented on lib/compilers/java.js:
```diff
@@ -0,0 +1,265 @@
+        return filteredOptions;
+    }
+
+    processAsm(result) {
```
Could some tests be added to ensure we don't unintentionally break the this processing function(s)? Otherwise, they look great!

mattgodbolt replied:
+1 to this! Love this PR but having some tests for the asm parsers has been beyond invaluable in ensuring the quality of Compiler Explorer's output.

RabsRincon commented on etc/config/java.defaults.properties:
```diff
@@ -0,0 +1,17 @@
+# Default settings for Java
```
We ought to think how we can get a hold of this for the live site, but this is a great start

mattgodbolt replied:
Looks like @partouf has done a great job on this already!

RabsRincon commented on etc/config/java.defaults.properties:
```diff
@@ -0,0 +1,17 @@
+options=
+supportsBinary=false
+needsMulti=false
+supportsExecute=false
```
Can you please ensure that there's a newline at EOL? :)

alshain replied:
Done :)

alshain commented:
Thanks for the comments so far.
And thanks for the Amazon support @partouf, I would have struggled with that.

I try to get in the changes from the comments later today or tomorrow!

In terms of tests, any preference which existing tests I should use as a reference?

RabsRincon commented:
You could take a look at `test/llvm-ir-parser-tests.js`, `test/demangler-tests.js` and `test/pascal-tests.js` as examples of languages checking that we don't break it (Plus how we check demangling in general) :)
Let us know if you need any help with the trickier parts!

alshain commented:
Okay, I've added a few test cases that the parsing of one or multiple asm files. I've also added a few test cases for the command line argument sanitization.

RabsRincon commented:
Then it's all set! We need to test this before merging, but it looks great! Thank you so much for taking the time to implement it and write those tests too :)

mattgodbolt commented:
Looks great! Thanks so much for this!

alshain commented:
Thanks to you, too!

mattgodbolt commented:
I've just been testing this in our staging environment and unfortunately have hit a snag: we "only" have 1.25GB available to run compilers (limited via `rlimit` on the address space). I've been unable to get `javac` to run.

By default I get:

```
Error occurred during initialization of VM
Could not allocate metaspace: 1073741824 bytes
Compiler returned: 1
```

With some hackery and `-J-XX:MaxHeapSize=750m -J-XX:+UseConcMarkSweepGC` stuff found by Googleing (this is the best outcome I could get) :

```
[0.066s][warning][gc] CMSMarkStack allocation failure
[0.066s][warning][gc] Failed to allocate CMS Marking Stack
Error occurred during initialization of VM
Could not create CMS collector
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
Compiler returned: 1````

mattgodbolt commented:
(any more than 750m and I got "not enough memory".

Unfortunately our current isolation technique (`firejail`) has a limitation where it only limits the address space. We'd ideally limit resident space instead, and that _might_ be better. We're looking at alternatives to `firejail` (including `nsjail`) which might help. Ideally though we'd find a way to make `javac` et al less memory hungry.

Any ideas welcomed!

mattgodbolt commented:
Have just tried setting `MALLOC_ARENA_MAX=4` as an environment variable...also no luck.

mattgodbolt commented:
The minimum I can get it to work with is ~3GB of address space. We might have to hack about with this unless we can get `nsjail` working... @apmorton I wonder if you've had a chance to look at that solution? :)

ThadHouse commented:
You have to use a mix of -Xms, -Xmn and -Xmx to get everything working. I've noticed sometimes if you only use one of them, the others will still override and use more virtual memory. This was a long time ago on 32 bit systems, but via the docs it should still work.

iakshay commented:
Would `UseCGroupMemoryLimitForHeap` help here? https://blogs.oracle.com/java/java-heap-sizing-in-a-container%3a-quickly-and-easily

valexey commented:
It seems like there is a solution here: https://github.com/compiler-explorer/compiler-explorer/issues/1438#issuecomment-693637417
